### ISSUE TITLE
Why does `IdentityLinearOperator` support rectangular structures

### ISSUE THREAD

**Issue by jpbrodrick89** [google/lineax#239]
@patrick-kidger  The documentation of `IdentityLinearOperator` states:

> If [`output_structure` is] passed then it must have the same number of elements as `input_structure`, so that the operator is square.

However, #50 added support for rectangular structures, this seems intentional as there is even a test that enforces this—`test_identity_with_different_structures` does not just check same size different layouts it explicitly tests different SIZES (5 and 7). What was the motivation for this change of behaviour? What would be the implication of removing it?

This leads to some strange behaviour:

- A rectangular `IdentityLinearOperator` will evaluate as `True` for `is_diagonal` and `is_tridiagonal` (no size matching guard)
- Solving ANY `IdentityLinearOperator` will return the original identity linear operator unchanged (rather than doing `op.T.mv(vector)` which would provide the correct solution).

These are all bugs that have obvious fixes but I'm wondering if we should also drop support for rectangular linear operator (and maybe add a permutation/pivot operator if we need the kind of functionality a rectangular identity provides).

**Comment by patrick-kidger** [comment]
I recall needing support for different *structures*. I no longer recall why we might need different *sizes*... which I agree seems a little strange.

**Comment by jpbrodrick89** [comment]
Thanks @patrick-kidger , I'll see if I can cleanly drop support for different sizes being sure we mark this as a breaking change. If we can think of a reason why we need different sizes before we merge dev into main and drop the next release (probably in the next 4-6 weeks I'm thinking) we can revert.

My current feel is permutation/pivot operator won't see much usage in lineax in the near term until we have better ways to exploit it (e.g. ComposedLinearSolver which I'm considering more in scope for lineaxtra) and I therefore plan to leave that out of the next release.

**Comment by patrick-kidger** [comment]
Makes sense! I'd double check in Optimistix/Diffrax in case anything downstream is already clearly depending on this.

(+Maybe also the commit that introduces this might say something more about why I put this in 😄)

**Comment by jpbrodrick89** [comment]
Checked them all, I think in the entire kidgerverse you only use `IdentityLinearOperator` once (for `EvalGradHessian` in `damped_newton_step`) and there it is square. Unfortunately the PR did not have any comment just "Various fixes for IdentityLinearOperator with different input/output structures", the only thing that might give a clue to your original intent is the commented out lines in the test, where there is a structure with a `None` leaf and `transpose` seems to accept a per-leaf boolean mask. It might have been something to do with the preceding commit #49 which "Fixed a crash with symbolic zero cotangent" which might have been the source of pytree's with missing leaves. All in all, it might have been necessary to workaround some bug when doing partial differentiation which feels like would be a JAX bug to me that is hopefully fixed.

```python
def test_identity_with_different_structures():
    structure1 = (
        jax.ShapeDtypeStruct((), jnp.float32),
        jax.ShapeDtypeStruct((2, 3), jnp.float16),
    )
    structure2 = {"a": jax.ShapeDtypeStruct((5,), jnp.float32)}
    # structure3 = (None, jax.ShapeDtypeStruct((2, 3), jnp.float16))
    op1 = lx.IdentityLinearOperator(structure1, structure2)
    op2 = lx.IdentityLinearOperator(structure2, structure1)
    # op3 = lx.IdentityLinearOperator(structure3, structure2)

    assert op1.T == op2
    # assert op2.transpose((True, False)) == op3
    assert jnp.array_equal(op1.as_matrix(), jnp.eye(5, 7))
    assert op1.in_size() == 7
    assert op1.out_size() == 5
    vec1 = (
        jnp.array(1.0, dtype=jnp.float32),
        jnp.array([[2, 3, 4], [5, 6, 7]], dtype=jnp.float16),
    )
    vec2 = {"a": jnp.array([1.0, 2.0, 3.0, 4.0, 5.0], dtype=jnp.float32)}
    vec1b = (
        jnp.array(1.0, dtype=jnp.float32),
        jnp.array([[2, 3, 4], [5, 0, 0]], dtype=jnp.float16),
    )
    assert shaped_allclose(op1.mv(vec1), vec2)
    assert shaped_allclose(op2.mv(vec2), vec1b)
```

**Comment by patrick-kidger** [comment]
How mysterious :D
Well past me clearly didn't do a good enough job documenting this, so I vote we fix this now, and blame him if something goes wrong!
